src/Build/Evaluation/Expander.cs

```diff
@@ -3268,6 +3275,122 @@ internal object Execute(object objectInstance, IPropertyProvider<T> properties,
                 }
             }
 
+            /// <summary>
+            /// Shortcut to avoid calling into binding if we recognize some most common functions.
+            /// Binding is expensive and throws first-chance MissingMethodExceptions, which is
+            /// bad for debugging experience and has a performance cost.
+            /// A typical binding operation with exception can take ~1.500 ms; this call is ~0.050 ms
+            /// (rough numbers just for comparison).
+            /// See https://github.com/Microsoft/msbuild/issues/2217
+            /// </summary>
+            /// <param name="objectInstance">Object that the function is called on</param>
+            /// <param name="args">arguments</param>
+            /// <returns>The value returned from the function call</returns>
+            private object ExecuteWellKnownFunction(object objectInstance, object[] args)
+            {
+                if (objectInstance is string text)
```

Comment:
Pattern matching is great for this method, but unfortunately it looks like it's causing the Mono CI builds to fail.  I'm not sure if we can update the Mono builds to support C# 7, or whether you should rewrite this without pattern matching.

Reply:
Yeah, I'll rewrite without C# 7.